### PR TITLE
Use tsync timestamps in NWBFile

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
@bendichter Alessio added the tsync timestamp capabilities into the extractors a while back, just now getting around to allowing the converter to actually use them via conversion options.

So far, the Recording itself, the accelerometer data, and the Image series interfaces all support handling this.

One interesting case with the image series is the camera frames actually go slightly (5 frames) beyond the final timestamp of the recording, so these are filled in with the unsychronized timestamps.

Pointing this to `add_multi_accelerometer` for readability.

### CODE REVIEW

**Comment by CodyCBakerPhD** [comment]
@alejoe91 Since our last discussion on this topic, I've actually decided the easiest way (i.e., least amount of code) to pass those `timestamps` is your first suggestion of the local override of `run_conversion` in this converter. It still operates on the `conversion_options` routing framework, however, which makes it a bit easier than the other things we were considering.

**Comment by CodyCBakerPhD** [comment]
@bendichter Ready for review; all timestamps are now being passed and used by downstream interfaces

Also handles electrode metadata, and polished the main conversion script a bit for subject and session info.

**Comment by CodyCBakerPhD** [comment]
@bendichter Updates per our last couple meetings

1. Adjusted main conversion script to work with new pipeline design.

2. Extended only the main recording and accelerometer data to include tsync timesteps.

3. Added static `quick_write` function for use in the quick save option of new pipeline; essentially a wrapper to the one-liner of `NwbSortingExtractor.write_sorting`, but since that is intended to potentially be the first time that NWBFile is written, it needs to be made with those minimal number of required `kwargs` such as description/id, so the `quick_write` function parses that the same automatic way it would during `get_metadata`, but independent of that sort of converter class use. So the call in the pipeline itself is still just one line

4. Also removed `dateparser` as requirement; using `strptime` instead.

**Comment by bendichter** [comment]
Thanks @CodyCBakerPhD. Is there an advantage to making `quick_write` a static method over just a regular function?

**Comment by CodyCBakerPhD** [comment]
@bendichter Hmm... not really; it would just come down to how to import it into the environment. If not static, then it would be an import like `from .syntalosnwbconverter import quick_write` or something like that... would that be preferable?

**Comment by bendichter** [comment]
@CodyCBakerPhD I think that would be better. Static classes are useful if 1) We want to enforce a common syntax across classes (as is the case for `RecordingExtractor.write_recording` or 2) if the method is only expected to be used within that class. If neither of those are true, it's probably better to make it a stand-alone method.

**Comment by CodyCBakerPhD** [comment]
@bendichter OK, made it a general method that just gets imported directly

File: mease_lab_to_nwb/convert_syntalos/syntaloseventinterface.py [hunk 1]
```diff
@@ -34,7 +35,7 @@ def run_conversion(self, nwbfile: NWBFile, metadata: dict):
         event_file = self.source_data['file_path']
         events_data = pd.read_csv(event_file, header=0)
         split_first_col = [x.split(";") for x in events_data['Time;Tag;Description']]
-        event_timestamps = [int(x[0]) for x in split_first_col]
+        event_timestamps = [int(x[0]) / 1E3 for x in split_first_col]
```
Comment (alejoe91) [inline mease_lab_to_nwb/convert_syntalos/syntaloseventinterface.py:38]: @CodyCBakerPhD a nicer way to do this is:

```
events_data = pd.read_csv(event_file, delimiter=';')
event_timestamps = events_data["Time"].to_numpy() / 1E3
event_labels = events_data["Tag"].to_numpy()
```

File: mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py [hunk 2]
```diff
@@ -36,17 +36,20 @@ def run_conversion(self, nwbfile: NWBFile, metadata: dict):
...
         video_file_path_list = [str(x.absolute()) for x in video_folder.iterdir() if x.suffix == ".mkv"]
 
-        video_timestamps = list()
+        video_timestamps = np.empty(0)
         for video_file_path in video_file_path_list:
             video_time_file = pd.read_csv(video_file_path.replace(".mkv", "_timestamps.csv"), header=0)
-            video_timestamps.append(np.array([int(x.split(";")[1]) for x in video_time_file['frame; timestamp']]))
+            video_timestamps = np.append(
+                video_timestamps,
+                [int(x.split(";")[1]) / 1E3 for x in video_time_file['frame; timestamp']]
```
Comment (alejoe91) [inline mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py:44]: Same here:

```
video_time_df = pd.read_csv(video_file_path.replace(".mkv", "_timestamps.csv"), delimiter=';')
# note that there is an additional space in timestamps
ts_column = [col for col in video_time_df.columns if 'timestamp' in col][0]
video_timestamps = video_time_df[ts_column].to_numpy() / 1E3
```

File: mease_lab_to_nwb/convert_syntalos/syntaloseventinterface.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntaloseventinterface.py]: Cleaned it up now, thanks for pointing out the delimiter usage here

File: mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py (hunk 2 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py]: Even further cleaned with the use of `skipininitialspace` arg in `read_csv`

File: mease_lab_to_nwb/convert_syntalos/convert_syntalos.py [hunk 3]
```diff
@@ -1,27 +1,52 @@
 """Authors: Alessio Buccino, Cody Baker, Szonja Weigl, and Ben Dichter."""
 from pathlib import Path
+from isodate import duration_isoformat
+from datetime import timedelta
 
-from .syntalosnwbconverter import SyntalosNWBConverter
+from mease_lab_to_nwb import SyntalosNWBConverter
```
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: use relative import

File: mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py [hunk 4]
```diff
@@ -23,30 +22,24 @@ def get_source_schema(cls):
...
-        """
-        Primary conversion function for the custom Syntalos image interface.
-
-        Parameters
-        ----------
-        nwbfile : NWBFile
-        metadata_dict : dict
-        stub_test : bool, optional
-            If true, truncates all data to a small size for fast testing. The default is False.
-        """
```
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py:35]: why did you delete this?

File: mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py [hunk 5]
```diff
@@ -14,7 +14,6 @@ class SyntalosImageInterface(BaseDataInterface):
 
     @classmethod
     def get_source_schema(cls):
-        """Return a partial JSON schema indicating the input arguments and their types."""
```
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py:17]: Why did you delete this docstring?

File: mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py [hunk 6]
```diff
@@ -19,22 +55,66 @@ class SyntalosNWBConverter(NWBConverter):
     )
 
     def get_metadata(self):
-        """Auto-populate as much metadata as possible."""
```
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py:22]: why delete the docstring?

File: mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py [hunk 7]
```diff
@@ -19,22 +55,66 @@ class SyntalosNWBConverter(NWBConverter):
...
+        metadata = super().get_metadata()
+        metadata['NWBFile'].update(
+            institution="EMBL - Heidelberg",
+            lab="Mease"
+        )
```
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py:62]: Why update metadata['NWBFile'] in two different places instead of all at once?

File: mease_lab_to_nwb/convert_syntalos/syntalosrecordinginterface.py [hunk 8]
```diff
@@ -30,38 +39,53 @@ def write_accelerometer_data(nwbfile: NWBFile, recording, stub_test: bool = Fals
...
 
     RX = SyntalosRecordingExtractor
 
-    def run_conversion(self, nwbfile: NWBFile, metadata: dict, stub_test: bool = False, add_accelerometer: bool = True):
+    def get_metadata(self):
+        intan_filepath = [x for x in Path(self.source_data['folder_path']).iterdir() if x.suffix == ".rhd"][0]
+        temp_intan_interface = IntanRecordingInterface(file_path=intan_filepath)
+        return temp_intan_interface.get_metadata()
+
+    def run_conversion(self, nwbfile: NWBFile, metadata: dict, stub_test: bool = False, add_accelerometer: bool = True,
```
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/syntalosrecordinginterface.py:87]: metadata is not named properly in docstring

**Review by bendichter — CHANGES_REQUESTED** [review]
a few small things

File: mease_lab_to_nwb/convert_syntalos/convert_syntalos.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: The notebook uses from mease_lab_to_nwb, shall I change it there as well?

File: mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py:35]: I realized the docstrings for the abstract functions are inherited  from the BaseDataInterfaces, so I'm getting into the habit of not duplicating them every time unless there are specific overrides to add extra arguments.

Comment (bendichter) [reply mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py:35]: oh, I see this feature of python was added in 3.5. Huh, I didn't know that! OK, carry on deleting, looks good to me.

File: mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py (hunk 6 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py:22]: Same reason as above, but I just realized the `NWBConverter .run_conversion` does indeed override a few things so I'm adding one there.

File: mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py (hunk 7 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntalosnwbconverter.py]: Good catch

File: mease_lab_to_nwb/convert_syntalos/syntalosrecordinginterface.py (hunk 8 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntalosrecordinginterface.py:87]: Fixed

File: mease_lab_to_nwb/convert_syntalos/convert_syntalos.py (hunk 3 above)
Comment (bendichter) [reply mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: ?

File: mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/syntalosimageinterface.py:35]: Yeah I had no clue either until I tried it - realizing now that we can cut down a lot of downstream code text by centralizing all of it on the nwb-conversion-tools side, which I'll work on polishing over the next week or two

File: mease_lab_to_nwb/convert_syntalos/convert_syntalos.py (hunk 3 above)
Comment (bendichter) [inline mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]:
```suggestion
from . import SyntalosNWBConverter
```

Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: The processing pipeline for Syntalos, uses `from mease_lab_to_nwb import SyntalosRecordingExtractor`. The ReadMe also specifies installing it as a package, so they should have one called `mease_lab_to_nwb`.

But if the primary script is to use a relative import, should the processing pipeline as well?

Also, the tank-lab-to-nwb uses `tank_lab_to_nwb` for the external script, but relative imports on other components. Should that be made relative as well?

Comment (CodyCBakerPhD) [reply mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: Also, I was able to get relative imports to work on my system, but only for downstream components where the parent package is known (i.e., all the other interfaces in this repo that use relative imports). But the main script doesn't associate with a parent package, so trying to use a relative import here returns an error.

Shall I assume it's something specific to my system, and change this back to a relative import here, or keep the version that works for me?

Comment (bendichter) [reply mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: All imports from within the same repo should be relative except for tests, which should always use absolute imports, and example code, like in a README file.

Comment (bendichter) [reply mease_lab_to_nwb/convert_syntalos/convert_syntalos.py:6]: I guess it's fine here, since this is being run a script.

Ideally there would be a command-line interface and this would be run as a package, and the relative import should work in that case
